src/DescriptorsCommand.cc

```diff
@@ -409,94 +412,240 @@ void AddDescriptor::retrieve_aws_descriptorSet(const std::string &set_path) {
+    //get reference tag for source node for ID
+    // Loop over properties list, add relevant query, link, and edges for each
+    for(int i=0; i < nr_expected_descs; i++) {
+        int node_ref = query.get_available_reference();
+        Json::Value cur_props;
+        cur_props = prop_list[i];
+        //TODO Note using iterator to modify ID return, we're gonna want to watch this closely.
+        cur_props[VDMS_DESC_ID_PROP] = Json::Int64(id+i);
+        cur_props[VDMS_DESC_LABEL_PROP] = label;
+
+        query.AddNode(node_ref, VDMS_DESC_TAG, cur_props, Json::nullValue);
+
+        // It passed the checker, so it exists.
+        int set_ref = query.get_available_reference();
+
+        Json::Value link;
+        Json::Value results;
+        Json::Value list_arr;
+        list_arr.append(VDMS_DESC_SET_PATH_PROP);
+        list_arr.append(VDMS_DESC_SET_DIM_PROP);
+        results["list"] = list_arr;
+
+        //constraints for getting set node to link to.
+        Json::Value constraints;
+        Json::Value name_arr;
+        name_arr.append("==");
+        name_arr.append(set_name);
+        constraints[VDMS_DESC_SET_NAME_PROP] = name_arr;
+
+        bool unique = true;
+
+        // Query set node-We only need to do this once, outside of the loop TODO MOVE
+        query.QueryNode(set_ref, VDMS_DESC_SET_TAG, link, constraints, results,
+                        unique);
+
+        //note this implicitly means that every node of a batch uses the same link
+        if (cmd.isMember("link")) {
+            add_link(query, cmd["link"], node_ref, VDMS_DESC_EDGE_TAG);
+        }
 
-  Json::Value constraints;
-  Json::Value name_arr;
-  name_arr.append("==");
-  name_arr.append(set_name);
-  constraints[VDMS_DESC_SET_NAME_PROP] = name_arr;
+        Json::Value props_edge;
+        query.AddEdge(-1, set_ref, node_ref, VDMS_DESC_SET_EDGE_TAG, props_edge);
+    }
 
-  bool unique = true;
+    return 0;
+}
 
-  // Query set node
-  query.QueryNode(set_ref, VDMS_DESC_SET_TAG, link, constraints, results,
-                  unique);
+int AddDescriptor::construct_protobuf(PMGDQuery &query,
+                                      const Json::Value &jsoncmd,
+                                      const std::string &blob, int grp_id,
+                                      Json::Value &error) {
 
-  if (cmd.isMember("link")) {
-    add_link(query, cmd["link"], node_ref, VDMS_DESC_EDGE_TAG);
+  bool batch_mode;
+  int rc;
+  const Json::Value &cmd = jsoncmd[_cmd_name];
+  const std::string set_name = cmd["set"].asString();
+
+
+  Json::Value prop_list = get_value<Json::Value>(cmd, "batch_properties");
+  if(prop_list.size() == 0){
+      rc = add_single_descriptor(query, jsoncmd, blob, grp_id, error);
+  } else {
+      rc = add_descriptor_batch(query, jsoncmd, blob, grp_id, error);
   }
 
-  Json::Value props_edge;
-  query.AddEdge(-1, set_ref, node_ref, VDMS_DESC_SET_EDGE_TAG, props_edge);
+  if(rc < 0) error["status"] = RSCommand::Error;
```

Comment:
More descriptive error please. we have the jsoncmd that failed. Any information to identify the failing command will be very helpful to a client.

Reply:
Let me check this, the message may be set elsewhewre, which is distinct from the status code.

Reply:
Okay, yeah, the actual error info string is set earlier in all the cases I found, so this should be okay, though I agree its a little weird. Ill see if I can refactor.